cmd/juju/controller/kill.go

```diff
@@ -122,15 +137,107 @@ func (c *killCommand) Run(ctx *cmd.Context) error {
-		ctx.Infof(fmtCtrStatus(ctrStatus))
-		for _, envStatus := range envsStatus {
-			ctx.Verbosef(fmtModelStatus(envStatus))
+	uuid := controllerEnviron.Config().UUID()
+	cloudName := controllerEnviron.Config().Type()
+	if err := c.WaitForModels(ctx, api, cloudName, uuid); err != nil {
+		c.DirectDestroyRemaining(ctx, api, cloudName)
+	}
+	return environs.Destroy(controllerName, controllerEnviron, store)
+}
+
+// DirectDestroyRemaining will attempt to directly destroy any remaining
+// models that have machines left.
+func (c *killCommand) DirectDestroyRemaining(ctx *cmd.Context, api destroyControllerAPI, cloudName string) {
+	hasErrors := false
+	hostedConfig, err := api.HostedModelConfigs()
+	if err != nil {
+		hasErrors = true
+		logger.Errorf("unable to retrieve hosted model config: %v", err)
+	}
+	for _, model := range hostedConfig {
+		ctx.Infof("Killing %s/%s via %s", model.Owner.Canonical(), model.Name, cloudName)
+		cfg, err := config.New(config.NoDefaults, model.Config)
+		if err != nil {
+			logger.Errorf(err.Error())
+			hasErrors = true
+			continue
+		}
+		env, err := environs.New(environs.OpenParams{
+			Cloud:  model.CloudSpec,
+			Config: cfg,
+		})
+		if err != nil {
+			logger.Errorf(err.Error())
+			hasErrors = true
+			continue
+		}
+		if err := env.Destroy(); err != nil {
+			logger.Errorf(err.Error())
+			hasErrors = true
+		} else {
+			ctx.Infof("  done")
+		}
+	}
+	if hasErrors {
+		logger.Errorf("there were problems destroying some models, manual intervention may be necessary to ensure resources are released")
+	} else {
+		ctx.Infof("All hosted models destroyed, cleaning up controller machines")
+	}
+}
+
+// WaitForModels will wait for the models to bring themselves down nicely.
+// It will return the UUIDs of any models that need to be removed forceably.
+func (c *killCommand) WaitForModels(ctx *cmd.Context, api destroyControllerAPI, cloudName, uuid string) error {
+	updateStatus := newTimedStatusUpdater(ctx, api, uuid, c.clock)
+
+	ctrStatus, modelsStatus := updateStatus(0)
+	lastStatus := ctrStatus
+	thirtySeconds := (time.Second * 30)
+	lastChange := c.clock.Now().Truncate(time.Second)
+	deadline := lastChange.Add(c.timeout)
+	for ; hasUnDeadModels(modelsStatus) &&
+		(deadline.After(c.clock.Now())); ctrStatus, modelsStatus = updateStatus(5 * time.Second) {
+		now := c.clock.Now().Truncate(time.Second)
+		if ctrStatus != lastStatus {
+			lastStatus = ctrStatus
+			lastChange = now
+			deadline = lastChange.Add(c.timeout)
+		}
+		timeSinceLastChange := now.Sub(lastChange)
+		timeUntilDestruction := deadline.Sub(now)
+		warning := ""
+		// We want to show the warning if it has been more than 30 seconds since
+		// the last change, or we are within 30 seconds of our timeout.
+		if timeSinceLastChange > thirtySeconds || timeUntilDestruction < thirtySeconds {
+			warning = fmt.Sprintf(" - will kill directly via %s in %s", cloudName, timeUntilDestruction)
```

Comment:
I having second thoughts about this. "will kill directly via google" is particularly bad/humorous - "google" is normally associated with the search engine.

How about if the full line looked something like:
```
Waiting on 1 model, 1 machine. Managed destruction nearing timeout - will kill machines directly in 25s
```